packages/editor/src/components/block-removal-warnings/index.js

```diff
@@ -0,0 +1,80 @@
+/**
+ * WordPress dependencies
+ */
+
+import { __ } from '@wordpress/i18n';
+import { privateApis as blockEditorPrivateApis } from '@wordpress/block-editor';
+import { useSelect } from '@wordpress/data';
+
+/**
+ * Internal dependencies
+ */
+import { unlock } from '../../lock-unlock';
+import { store as editorStore } from '../../store';
+
+const { BlockRemovalWarningModal } = unlock( blockEditorPrivateApis );
+
+// Prevent accidental removal of certain blocks, asking the user for confirmation first.
+const blockRemovalRules = {
+	'core-query': {
+		message: __(
+			'The Query Loop block displays a list of posts or pages, so removing it will prevent that content from displaying.'
+		),
+		postTypes: [ 'wp_template' ],
+		callback: ( blockName, ruleKeysForPrompt ) => {
+			if ( blockName === 'core/query' ) {
+				ruleKeysForPrompt.add( 'core-query' );
+			}
+		},
+	},
```

Comment:
@talldan I have revisited the rules to try and make them more generic, and also to add in the ability to limit the messages by post type. 

So now every rule needs:

key - this is the property name and is used to map to the messages in the confirmation prompt
    message - message to display
    postTypes - array of post types on which to display the message
    callback - method to check if warning should show - callback gets `blockName`, `ruleKeysForPrompt` and `blockAttributes` passed in

Doing it this way makes for a bit of repetition for simple block checks, but the advantage of just havng a callback for each is devs are then free to calculate this however they want either for specific blocks or more generic like the patterns.

This is just a first stab at this bigger refactor of this API so happy to take thoughts on alternatives.

Reply:
we can probably make the individual block callback more generic and share it between the rules somehow if we decide to go  this way.